### ISSUE TITLE
Mac m1 Http works but https is broken?

### ISSUE THREAD

**Issue by helletheone** [nginx-proxy/nginx-proxy#1673]
Hi,

on http is everything it fine. But on https i got this:

```
nginx.1     | 2021/06/20 18:51:26 [notice] 30#30: signal 17 (SIGCHLD) received from 99
nginx.1     | 2021/06/20 18:51:26 [alert] 30#30: worker process 99 exited on signal 11
nginx.1     | 2021/06/20 18:51:26 [notice] 30#30: start worker process 103
nginx.1     | 2021/06/20 18:51:26 [notice] 30#30: signal 29 (SIGIO) received
nginx.1     | 2021/06/20 18:51:26 [notice] 30#30: signal 17 (SIGCHLD) received from 100
nginx.1     | 2021/06/20 18:51:26 [alert] 30#30: worker process 100 exited on signal 11
nginx.1     | 2021/06/20 18:51:26 [notice] 30#30: start worker process 104
nginx.1     | 2021/06/20 18:51:26 [notice] 30#30: signal 29 (SIGIO) received
nginx.1     | 2021/06/20 18:51:27 [notice] 30#30: signal 17 (SIGCHLD) received from 101
nginx.1     | 2021/06/20 18:51:27 [alert] 30#30: worker process 101 exited on signal 11
nginx.1     | 2021/06/20 18:51:27 [notice] 30#30: start worker process 105
nginx.1     | 2021/06/20 18:51:27 [notice] 30#30: signal 29 (SIGIO) received
nginx.1     | 2021/06/20 18:51:27 [notice] 30#30: signal 17 (SIGCHLD) received from 102
nginx.1     | 2021/06/20 18:51:27 [alert] 30#30: worker process 102 exited on signal 11
nginx.1     | 2021/06/20 18:51:27 [notice] 30#30: start worker process 106
nginx.1     | 2021/06/20 18:51:27 [notice] 30#30: signal 29 (SIGIO) received
nginx.1     | 2021/06/20 18:51:27 [notice] 30#30: signal 17 (SIGCHLD) received from 103
nginx.1     | 2021/06/20 18:51:27 [alert] 30#30: worker process 103 exited on signal 11
nginx.1     | 2021/06/20 18:51:27 [notice] 30#30: start worker process 107
nginx.1     | 2021/06/20 18:51:27 [notice] 30#30: signal 29 (SIGIO) received
nginx.1     | 2021/06/20 18:51:28 [notice] 30#30: signal 17 (SIGCHLD) received from 104
nginx.1     | 2021/06/20 18:51:28 [alert] 30#30: worker process 104 exited on signal 11
nginx.1     | 2021/06/20 18:51:28 [notice] 30#30: start worker process 108
nginx.1     | 2021/06/20 18:51:28 [notice] 30#30: signal 29 (SIGIO) received
nginx.1     | 2021/06/20 18:51:28 [notice] 30#30: signal 17 (SIGCHLD) received from 105
nginx.1     | 2021/06/20 18:51:28 [alert] 30#30: worker process 105 exited on signal 11
nginx.1     | 2021/06/20 18:51:28 [notice] 30#30: start worker process 109
nginx.1     | 2021/06/20 18:51:28 [notice] 30#30: signal 29 (SIGIO) received
```

here is the docker-compose.yml

```yaml
version: '3'
services:
  #######################################
  # Delegate requests to container via
  # Nginx reverse proxy
  #######################################
  nginx-proxy:
    container_name: nginx-proxy
    image: jwilder/nginx-proxy
    build:
      context: docker/nginx-proxy
    ports:
      - "80:80"
      - "443:443"
    networks:
      - default
    volumes:
      - /var/run/docker.sock:/tmp/docker.sock:ro
      - ./docker/nginx-proxy/nginx.tmpl:/app/nginx.tmpl:ro
      - ./docker/nginx-proxy/certs/default.key:/etc/nginx/certs/default.key:ro
      - ./docker/nginx-proxy/certs/default.crt:/etc/nginx/certs/default.crt:ro

networks:
  default:
    external:
      name: kult-dev-network
```

**Comment by buchdag** [comment]
Hi. Do you have the same issue with the upstream image (ie not locally built and no custom nginx.tmpl) ?

**Comment by lawmillion** [comment]
I have the same problem, how to solve it

**Comment by buchdag** [comment]
@lawmillion with the base image ?

Could you share your setup ?

**Comment by kblcuk** [comment]
@buchdag I've been playing with this a bit locally, and it seems to be that the culpit is in additional network, and not https. You can reproduce it with this `docker-compose.yml`: 

```
  services:
    nginx-proxy:
      image: nginxproxy/nginx-proxy
      ports:
        - "80:80"
      volumes:
        - /var/run/docker.sock:/tmp/docker.sock:ro
      networks:
        - proxynet

    whoami:
      image: jwilder/whoami
      expose:
        - "8000"
      environment:
        - VIRTUAL_HOST=whoami.local
        - VIRTUAL_PORT=8000

  networks:
    proxynet:
      name: proxynet
```
and adding a 
```
127.0.0.1 whoami.local
::1 whoami.local
```
to `/etc/hosts` file.

Then, running the whole shebang with `docker-compose up` and trying to `curl http://whoami.local` gives me:

```
nginx-proxy_1  | nginx.1     | 2021/06/30 14:14:35 [error] 57#57: *2 no live upstreams while connecting to upstream, client: 172.25.0.1, server: whoami.local, request: "GET / HTTP/1.1", upstream: "http://whoami.local-upstream/", host: "whoami.local"
nginx-proxy_1  | nginx.1     | whoami.local 172.25.0.1 - - [30/Jun/2021:14:14:35 +0000] "GET / HTTP/1.1" 502 157 "-" "curl/7.64.1" "whoami.local-upstream"
```

However if I comment out the `networks:` part from `ngingx-proxy` service, everything works as expected, and `curl http://whoami.local` results in:

```
whoami_1       | I'm 7ebc8d931530
nginx-proxy_1  | nginx.1     | whoami.local 172.26.0.1 - - [30/Jun/2021:14:15:50 +0000] "GET / HTTP/1.1" 200 17 "-" "curl/7.64.1" "172.26.0.2:8000"
```

Let me know if I can help debugging this further.

**Comment by buchdag** [comment]
Could you give me the result of `docker exec nameofnginxproxycontainer nginx -T` in both cases ?

**Comment by kblcuk** [comment]
@buchdag 

They were almost identical, except for this bit:
```
❯ diff -u nginx-conf-test-with-network nginx-conf-test-no-network
--- nginx-conf-test-with-network        2021-07-01 11:10:41.000000000 +0300
+++ nginx-conf-test-no-network  2021-07-01 11:11:18.000000000 +0300
@@ -193,9 +193,9 @@
 }
 # whoami.local
 upstream whoami.local-upstream {
-       # Cannot connect to network 'proxy-playground_default' of this container
-       # Fallback entry
-       server 127.0.0.1 down;
+       ## Can be connected with "proxy-playground_default" network
+       # proxy-playground_whoami_1
+       server 172.26.0.2:8000;
 }
 server {
        server_name whoami.local;
```

**Comment by buchdag** [comment]
Okay looks like I was not paying enough attention, in your test the `jwilder/whoami` container also need to be on the `proxynet` network.

**Comment by kblcuk** [comment]
🤦 well this is embarrassing. 

Now that I added the missing `proxynet` to whoami container, it works both with http and https! Except... it still fails if I use some other image than `jwilden/whoami`!

I pushed simplest example I could come up with to this repo: https://github.com/kblcuk/dev-proxy-1673-demo. I tried it with node images (since for me locally those are the ones that fail), and it doesn't seem to matter that much whether it's alpine flavor or not. I also tried this [alpine-based python image](https://github.com/jfloff/alpine-python), the result is the same -- no body in server responce (`curl: (52) Empty reply from server`), and
```
nginx-proxy_1  | nginx.1     | 2021/07/01 12:42:18 [notice] 27#27: signal 17 (SIGCHLD) received from 45
nginx-proxy_1  | nginx.1     | 2021/07/01 12:42:18 [alert] 27#27: worker process 45 exited on signal 11
nginx-proxy_1  | nginx.1     | 2021/07/01 12:42:18 [notice] 27#27: start worker process 48
nginx-proxy_1  | nginx.1     | 2021/07/01 12:42:18 [notice] 27#27: signal 29 (SIGIO) received
```
in container log.

**Comment by buchdag** [comment]
Your base example contains way too much extra config / noise IMHO. You should try something more in the line of:

  
```yaml
services:
  nginx-proxy:
    image: nginxproxy/nginx-proxy
    ports:
      - "443:443"
      - "80:80"
    volumes:
      - /var/run/docker.sock:/tmp/docker.sock:ro
      - ./certs:/etc/nginx/certs

  app:
    image: nginx
    environment:
      - VIRTUAL_HOST=app.local
```

No need to specify a network as in this setup compose will automatically create one and connect every containers to it.

Same thing for the `expose` directive on the nginx container, [the Dockerfile already expose this port](https://github.com/nginxinc/docker-nginx/blob/f3fe494531f9b157d9c09ba509e412dace54cd4f/mainline/debian/Dockerfile#L112).

Put `app.local` in your `/etc/hosts` or test using either

```
curl -k --header 'Host: app.local' https://127.0.0.1
```

or

```
curl -k --resolve app.local:443:127.0.0.1 https://app.local/
``` 

then build up from there.

**Comment by kblcuk** [comment]
@buchdag yea, that looks somewhat simpler 😄 However result is still the same, running your example as is (together with cert generated, naturally) gives me

```
❯ curl -k --header 'Host: app.local' https://127.0.0.1
curl: (52) Empty reply from server
```
and
```
nginx-proxy_1  | nginx.1     | 2021/07/01 13:30:38 [notice] 30#30: signal 17 (SIGCHLD) received from 54
nginx-proxy_1  | nginx.1     | 2021/07/01 13:30:38 [alert] 30#30: worker process 54 exited on signal 11
nginx-proxy_1  | nginx.1     | 2021/07/01 13:30:38 [notice] 30#30: start worker process 57
nginx-proxy_1  | nginx.1     | 2021/07/01 13:30:38 [notice] 30#30: signal 29 (SIGIO) received
```
from container. Any other debug info I can provide?

**Comment by buchdag** [comment]
> Any other debug info I can provide?

I think next step would be running the three container setup with nginx in debug mode. Do you know how to do this or do you need assistance ?

**Comment by kblcuk** [comment]
I followed [separate containers section](https://github.com/nginx-proxy/nginx-proxy#separate-containers) of readme, and got everything running. 
The resulting `docker-compose.yml` looks like this:
```
services:
  nginx:
    image: nginx
    container_name: nginx
    ports:
      - "443:443"
      - "80:80"
    volumes:
      - /etc/nginx/conf.d
      - ./certs:/etc/nginx/certs

  dockergen:
    image: jwilder/docker-gen
    command: -notify-sighup nginx -watch /etc/docker-gen/templates/nginx.tmpl /etc/nginx/conf.d/default.conf
    volumes_from:
      - nginx
    volumes:
      - /var/run/docker.sock:/tmp/docker.sock:ro
      - ./nginx.tmpl:/etc/docker-gen/templates/nginx.tmpl

  app:
    image: nginx
    environment:
      - VIRTUAL_HOST=app.local
```


The effect is still the same 😕 `nginx` container gets `SIGCHLD`, and we get no body in response.

**Comment by kblcuk** [comment]
Hmm, I wonder if this is actually something `curl`-specific? I've now tested what would happen if I use `wget` instead, and it passes just fine, without crashing nginx process:
```
❯ wget --no-check-certificate --header 'Host: app.local' --server-response -O /dev/null https://127.0.0.1
--2021-07-01 18:25:39--  https://127.0.0.1/
Connecting to 127.0.0.1:443... connected.
...
HTTP request sent, awaiting response...
  HTTP/1.1 200 OK
  Server: nginx/1.21.0
  Date: Thu, 01 Jul 2021 15:25:39 GMT
  Content-Type: text/html
  Content-Length: 612
  Connection: keep-alive
  Last-Modified: Tue, 25 May 2021 12:28:56 GMT
  ETag: "60aced88-264"
  Accept-Ranges: bytes
  Strict-Transport-Security: max-age=31536000
Length: 612 [text/html]
```

Running `curl` command with more verbose output shows me this:
```
❯ curl -k --header 'Host: app.local' https://127.0.0.1 -vvv
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
.... cert info ....
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x13500d200)
> GET / HTTP/2
> Host: app.local
> User-Agent: curl/7.64.1
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
* Empty reply from server
* Connection #0 to host 127.0.0.1 left intact
curl: (52) Empty reply from server
* Closing connection 0
```

**Comment by buchdag** [comment]
An issue with `HTTP/2` maybe ?

Could you try `curl -k --http1.1 --header 'Host: app.local' https://127.0.0.1 -vvv` to force HTTP/1.1 ?

To run the nginx container in debug mode, use this service definition instead :

```yaml
  nginx:
    image: nginx
    container_name: nginx
    ports:
      - "443:443"
      - "80:80"
    volumes:
      - /etc/nginx/conf.d
      - ./certs:/etc/nginx/certs
    command: [nginx-debug, '-g', 'daemon off;']
```

Maybe that'll give more information about why the nginx process crashes.

**Comment by kblcuk** [comment]
tl;dr -- a possible work-around would be to use [`nginxproxy/nginx-proxy:alpine` image](https://github.com/nginx-proxy/nginx-proxy#nginxproxynginx-proxyalpine) instead of a debian-based one.

---


@buchdag sorry for the late reply, was busy week 😅 

Yea, I thought about http1.1 as well, however forcing that doesn't seem to change anything. And nginx debug mode doesn't seem to log anything extra.

```
❯ curl -k --http1.1 --header 'Host: app.local' https://127.0.0.1 -vvv
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 443 (#0)
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-CHACHA20-POLY1305
* ALPN, server accepted to use http/1.1
* Server certificate:
* <redacted cert info> 
*  SSL certificate verify ok.
> GET / HTTP/1.1
> Host: app.local
> User-Agent: curl/7.64.1
> Accept: */*
>
* Empty reply from server
* Connection #0 to host 127.0.0.1 left intact
curl: (52) Empty reply from server
* Closing connection 0
```

Now the interesting bit is that this doesn't manifest if I use alpine flavour of nginx image (i.o.w. `nginx:alpine`)! The output looks quite close to the "broken" one:
```
❯ curl -k --header 'Host: app.local' https://127.0.0.1 -vvv
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-CHACHA20-POLY1305
* ALPN, server accepted to use h2
* Server certificate:
*  subject: O=mkcert development certificate; OU=alex@vackerflicka (Alexei Mikhailov)
*  start date: Jul  1 13:28:53 2021 GMT
*  expire date: Oct  1 13:28:53 2023 GMT
*  issuer: O=mkcert development CA; OU=alex@vackerflicka (Alexei Mikhailov); CN=mkcert alex@vackerflicka (Alexei Mikhailov)
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x12280d600)
> GET / HTTP/2
> Host: app.local
> User-Agent: curl/7.64.1
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
< HTTP/2 200
< server: nginx/1.21.0
< date: Mon, 12 Jul 2021 08:59:21 GMT
< content-type: text/html
< content-length: 612
< last-modified: Tue, 06 Jul 2021 14:59:17 GMT
< etag: "60e46fc5-264"
< accept-ranges: bytes
< strict-transport-security: max-age=31536000
<
<!DOCTYPE html>
<...redacted standard nginx page>
* Connection #0 to host 127.0.0.1 left intact
* Closing connection 0
```

While researching this I've stumbled upon a rather old, but very similar issue [in nginx issue tracker](https://trac.nginx.org/nginx/ticket/912), that claims the problem to happen due to `-O2` optimisation flags.

Now, looking at the `nginx -V` output from docker image, it seems that it used `-O2` during compilation, so that might be what causing this issue? And since alpine builds don't use gcc, this doesn't manifest?

```
❯ docker run nginx nginx -V
...
nginx version: nginx/1.21.1
built by gcc 8.3.0 (Debian 8.3.0-6)
built with OpenSSL 1.1.1d  10 Sep 2019
TLS SNI support enabled
configure arguments: --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --modules-path=/usr/lib/nginx/modules --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp --http-scgi-temp-path=/var/cache/nginx/scgi_temp --user=nginx --group=nginx --with-compat --with-file-aio --with-threads --with-http_addition_module --with-http_auth_request_module --with-http_dav_module --with-http_flv_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_mp4_module --with-http_random_index_module --with-http_realip_module --with-http_secure_link_module --with-http_slice_module --with-http_ssl_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --with-mail --with-mail_ssl_module --with-stream --with-stream_realip_module --with-stream_ssl_module --with-stream_ssl_preread_module --with-cc-opt='-g -O2 -fdebug-prefix-map=/data/builder/debuild/nginx-1.21.1/debian/debuild-base/nginx-1.21.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fPIC' --with-ld-opt='-Wl,-z,relro -Wl,-z,now -Wl,--as-needed -pie'
```

Anyhoo to sum up, using alpine-based nginx-proxy build instead of debian one solves the issue for my local env, which is good enough for me 😅 

Feel free to ping me if you want to test something!

**Comment by buchdag** [comment]
@kblcuk thanks for investigating this, seems there isn't much we can do from this project end. 😒 

At least the alpine based image is working. 👍

**Comment by buchdag** [comment]
Inactive issue, closing.
